ostd/src/mm/io.rs

```diff
@@ -389,12 +312,45 @@ impl<'a> VmReader<'a, KernelSpace> {
         }
     }
 
+    /// Reads all data into the writer until one of the three conditions is met:
+    /// 1. The reader has no remaining data.
+    /// 2. The writer has no available space.
+    /// 3. The reader/writer encounters some error.
+    ///
+    /// On success, the number of bytes read is returned;
+    /// On error, both the error and the number of bytes read so far are returned.
+    pub fn read_fallible(
+        &mut self,
+        writer: &mut VmWriter<'_, NonFallible>,
```

Comment:
I think the `read_fallible` implemented for `VmReader<Fallible>` should be able to take both `VmWriter<Fallible>` and `VmWriter<NonFallible>` as the target. It will be confused for users if a fallible reader can not read from a fallible writer. 

The `read_fallible` implemented for `VmReader<NonFallible>` can only take  `VmWriter<Fallible>` as target.

Reply:
Make sense

Reply:
Then this would violate the safety conditions of `memcpy_fallible`, as it claims:
```rust
/// # Safety
/// 
/// - Users should ensure one of [`src`..`src` + len] and [`dst`..`dst` + len]
///   is in user space, and the other virtual memory range is in kernel space
///   and is ensured to be [valid].
```

I have been thinking for some time that the safety conditions are not clear enough. Maybe we can try something like https://github.com/lrh2000/asterinas/commit/bbefffee66119e5d1a9e8a2e5d9a58b879537b95, and I think it seems to be much better.